include/roboteam_utils/containers/state_machine.hpp

```diff
@@ -0,0 +1,213 @@
+
+        /**
+         * Assertions that ensures that all the member functions are present
+         */
+        static_assert(type_traits::is_valid_type_v<Base, Enum, SI>, "Type Base does not have the required functions (noexcept): Enum update(SI const&), void initialize(), void terminate()");
+
+        /**
+         * STL-style using's
+         */
+        using iterator = typename decltype(_data)::iterator;
+        using const_iterator = typename decltype(_data)::const_iterator;
+        using size_type = typename decltype(_data)::size_type;
+
+    public:
+        /**
+         * Constructor that takes any amount of arguments, all of which must inherit from Base
+         *
+         * ~~~{cpp}
+         * struct Base {}; // imagine proper members
+         * struct Derived : public Base {};
+         * struct AnotherDerived : public Derived{};
+         *
+         * enum EnumType{}; // again, imagine proper members
+         *
+         * enum SkillInfo{};
+         *
+         * state_machine<Base, EnumType, SkillInfo>{ Base(), Derived(), AnotherDerived() };
+         * ~~~
+         *
+         * @tparam Tys Types of arguments
+         * @param args Argument pack
+         */
+        template <typename...Tys>
+        explicit state_machine(Tys&&... args) noexcept {
+            // sizeof... gets the amount of elements in a pack
+            // https://en.cppreference.com/w/cpp/language/sizeof...
+            _data.reserve(sizeof...(args));
+            // Fold expression
+            /**
+             * Fold expression, essentially:
+             * for element in args:
+             *  data.emplace_back(make_unique(ensure_reference_type(args))
+             *
+             * std::forward is a conditional cast to r-value reference, it's like std::move()
+             * but it won't std::move if it wasn't already std::move'd into the constructor
+             *
+             * You could for example print an argument pack by doing the following
+             *
+             * template <typename...Tys>
+             * void print_all(Tys... args) { // Copy for clarity
+             *      ((std::cout << args), ...);
+             * }
+             *
+             * Now say you'd want to add a space:
+             * ((std::cout << args << " "), ...);
+             *
+             * Add up a pack of ints together?
+             *
+             * int result = (args + ...);
+             */
+            (_data.emplace_back(std::make_unique<Tys>(std::forward<Tys>(args))), ...);
+        }
+
+        /**
+         * Gets a mutable iterator to the first element
+         * @return std::vector<Base>::iterator
+         */
+        iterator begin() noexcept {
+            return _data.begin();
+        }
+
+        /**
+         * Gets a mutable iterator to 1 past the last element
```

Comment:
Why 1 past?

Reply:
standard, 1 past so you can `for auto const& each : x`